### PR TITLE
Reafactor concurrent store

### PR DESCRIPTION
**Pull request by thejustinwalsh**
Using https://github.com/thejustinwalsh/react-concurrent-store to implement asset loading hooks that integrate with the pixi.js Asset singleton that is responsible for loading, caching, and processing assets for pixi.js applications.

Controlling the timing of cache actions like clearing, resetting, or reloading assets when recovering from an error boundary is challenging to implement with useSyncExternalStore ([known caveats](https://react.dev/reference/react/useSyncExternalStore#caveats)), and I was never able to perfectly time reloading an asset from the same cache key previously.

This implementation is passing all the previous tests in addition to the test that I was unable to get passing with useSyncExternalStore. I've also abandoned react internals to implement this feature and this is now production ready, pending publishing to npm.

I still have a few more tests to run through, namely using the non-suspense versions of the hooks within a suspense boundary, as previously reported by early users, as well as SSR / RSC testing.

Accompanying WIP early demo of consuming the API (under active iteration) is available here: https://stackblitz.com/edit/vitejs-vite-fg32zekh?file=src%2FApp.tsx

### CODE REVIEW

File: src/hooks/useAssetCache.ts [hunk 1]
```diff
@@ -1,88 +1,141 @@
...
-  let cached = cache.get(k);
-  if (!cached) {
-    cached = load();
-    cache.set(k, cached);
+const store = createStore<AssetCache, AssetCacheAction>(new Map(), storeReducer);
+
+export function unsafeClearCache() {
+  // @ts-expect-error
+  store._current = store._sync = store._transition = new Map();
+}
```
[the remark this answers is not in the document]
Comment (thejustinwalsh) [reply src/hooks/useAssetCache.ts:129]: Internal use only in tests; added comment to clarify.

File: src/hooks/useAssetCache.ts [hunk 2]
```diff
@@ -1,88 +1,141 @@
...
+    case 'remove': {
+      remove(action.payload.urls);
+      const key = createKey(action.payload.urls);
+      const hash = hashKey(key);
+      current.delete(hash);
+      return new Map(current);
+    }
+    // set: Direct cache mutation, skip re-render
+    case 'set': {
+      return current.set(action.payload.hash, action.payload);
```
[the remark this answers is not in the document]
Comment (thejustinwalsh) [reply src/hooks/useAssetCache.ts:104]: Adding clarifying comments, by design.

File: src/useAssetBundle.ts [hunk 3]
```diff
@@ -1,39 +1,68 @@
...
+    if (cache.data === null && !sync.current.has(cache.promise)) {
+      cache.promise
+        .then(data => {
+          if (sync.current.has(cache.promise)) {
+            setState({
+              data,
+              error: null,
+            });
+          }
+          return data;
```
[the remark this answers is not in the document]
Comment (thejustinwalsh) [reply src/useAssetBundle.ts:32]: @Copilot If I do not return a value from then, won't the promise fulfill to undefined, and any further chaining off this promise will have a value of undefined?
